packages/cozy-harvest-lib/src/components/TriggerLauncher.jsx

```diff
@@ -0,0 +1,150 @@
+import React, { Component } from 'react'
+import PropTypes from 'prop-types'
+
+import { withMutations } from 'cozy-client'
+
+import TwoFAForm from './TwoFAForm'
+import { accountsMutations } from '../connections/accounts'
+import { triggersMutations } from '../connections/triggers'
+
+import KonnectorJob from '../models/KonnectorJob'
+
+// Statuses
+const ERRORED = 'ERRORED'
+const IDLE = 'IDLE'
+const LOGIN_SUCCESS = 'LOGIN_SUCCESS'
+const TWO_FA_REQUEST = 'TWO_FA_REQUEST'
+const TWO_FA_MISMATCH = 'TWO_FA_MISMATCH'
+const PENDING = 'PENDING'
+const SUCCESS = 'SUCCESS'
+const RUNNING_TWOFA = 'RUNNING_TWOFA'
+
+/**
+ * Trigger Launcher renders its children with following props:
+ * * launch: Callback to launch the trigger
+ * * running: Boolean which indicates if the trigger is running
+ *
+ * ### Example
+ * ```js
+ * <TriggerLauncher trigger={trigger}>
+ *   {(launch, running) => (
+ *     <Button onClick={launch} disabled={running}
+ *    )}
+ * </TriggerLauncher>
+ * ```
+ *
+ * TODO: inject other props to indicates status (for example: loginSucceed),
+ * and use callbacks like `onLoginSuccess` to make this component usable in
+ * TriggerManager for example
+ */
+export class TriggerLauncher extends Component {
+  state = {
+    status: IDLE
+  }
+
+  constructor(props, context) {
+    super(props, context)
+
+    this.launch = this.launch.bind(this)
+    this.handleTwoFAModalDismiss = this.handleTwoFAModalDismiss.bind(this)
+    this.handle2FACode = this.handle2FACode.bind(this)
+    this.handleError = this.handleError.bind(this)
+    this.handleSuccess = this.handleSuccess.bind(this)
+    this.handleLoginSuccess = this.handleLoginSuccess.bind(this)
+    this.handleTwoFARequest = this.handleTwoFARequest.bind(this)
+  }
+
+  handleTwoFAModalDismiss() {
+    // TODO: Make the modale not closable, or offer possibility to re-open it.
+    this.setState({ showTwoFAModal: false })
+  }
+
+  handleError() {
+    this.setState({ status: ERRORED, showTwoFAModal: false })
+  }
+
+  handleSuccess() {
+    this.setState({ status: SUCCESS, showTwoFAModal: false })
+  }
+
+  handleLoginSuccess() {
+    this.setState({ status: LOGIN_SUCCESS })
+  }
+
+  handleTwoFARequest() {
+    this.setState({ status: TWO_FA_REQUEST, showTwoFAModal: true })
+  }
+
+  handleTwoFAMismatch() {
+    this.setState({ status: TWO_FA_MISMATCH, showTwoFAModal: true })
+  }
+
+  handle2FACode(code) {
+    this.setState({ status: RUNNING_TWOFA })
+    this.konnectorJob.sendTwoFACode(code)
+  }
+
+  async launch() {
+    const { trigger } = this.props
+    const { client } = this.context
+
+    this.setState({ status: PENDING })
+
+    this.konnectorJob = new KonnectorJob(client, trigger)
+    this.konnectorJob
+      .on('error', this.handleError)
+      .on('loginSuccess', this.handleLoginSuccess)
+      .on('success', this.handleSuccess)
+      .on('twoFARequest', this.handleTwoFARequest)
+      .on('twoFAMismatch', this.handleTwoFAMismatch)
+    this.konnectorJob.launch()
```

Comment:
Observations:

- In launch we are assigning this.konnectorJob, this means that TriggerLauncher has an internal state and it is not cleaned up on success or error. Do we need to save an internal variable ? Could we simplify and only use a closure ?
 - The component is responsible to remember the status of the konnector in its internal state. I wonder if it would no be simpler to have this status in the konnectorJob and rely on it for the render.

Reply:
Seems clever. The konnectorJob is saved as an interanl variable to have access to `konnectorJob.sendTwoFACode` method to pass it to TwoFAModal.

Keeping the status in konnectorJob could also spare us all the `handle{KonnectorJobEvent}` methods here.

Reply:
Agree about the status but we still need the internal variable for the `sendTwoFACode` prop passed to the form, I don't see another simpler way to handle it :/

Reply:
See https://github.com/cozy/cozy-libs/pull/415/commits/fee5d4745535bc71fb6badaf248b2b3fcfb847d3